patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,758 @@
++import com.mojang.brigadier.arguments.ArgumentType;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++
++/**
++ * An argument type that wraps around another argument.
++ * This argument is special in that the argument type that is wrapped will
++ * be sent to the client, not the wrapper itself.
++ *
++ * @param <T> type
++ * @param <B> base wrapping type
++ */
++@ApiStatus.Experimental
++public abstract class WrapperArgumentType<T, B> implements ArgumentType<T> {
++
++    @NotNull
++    private final ArgumentType<B> vanillaPrimitiveType;
++
++    public WrapperArgumentType(@NotNull ArgumentType<B> primitive) {
++        this.vanillaPrimitiveType = primitive;
++    }
++
++    /**
++     * Parses the argument using the wrapped argument type then converting
++     * it.
++     * @param reader string reader
++     * @return value
++     * @throws CommandSyntaxException if an error occurs while parsing
++     */
++    @NotNull
++    @Override
++    public T parse(@NotNull StringReader reader) throws CommandSyntaxException {
++        return this.convert(this.vanillaPrimitiveType.parse(reader));
++    }
++
++    /**
++     * Converts the type from the wrapped argument to this argument.
++     *
++     * @param baseType wrapped argument provided value
++     * @return converted value
++     * @throws CommandSyntaxException if an exception occurs while parsing
++     */
++    @NotNull
++    public abstract T convert(@NotNull B baseType) throws CommandSyntaxException;
++
++    /**
++     * Gets the wrapped argument.
++     *
++     * @return wrapped
++     */
++    @NotNull
++    public ArgumentType<B> getWrapped() {
++        return this.vanillaPrimitiveType;
++    }
++
++    /**
++     * Cannot be controlled by the server.
++     * Returned in cases where there are multiple arguments in the same node.
++     * This helps differentiate and tell the player what the possible inputs are.
++     *
++     * @return client set examples
++     */
++    @NotNull
++    @Override
++    public final Collection<String> getExamples() {
++        return this.vanillaPrimitiveType.getExamples();
++    }
++
++    /**
++     * Returns if the argument itself should handle sugggestion
++     * parsing instead of the wrapper.
++     *
++     * @return handles suggestions
++     */
++    public boolean handleSuggestions() {
++        return false;
++    }
++
++
++}
+diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
+index 15a9993c2705e2dcc0366394ded57acffab589c0..d8c05f5b931c8f7b26c6cc50dab3df335e76c32e 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -2334,6 +2334,7 @@ public final class Bukkit {
+      * @return the default message
+      */
+     @NotNull
++    @Deprecated // Used only for bukkit commands
```

Comment:
We are still supporting bukkit commands, no? Like CommandExecutor, extending Command, etc. I think this should be left as is, but the docs can change. Or maybe keep the deprecation, but have a new method? Idk, cause like its still useful, just not for all commands.

Reply:
Well yes, but it's essentially obsolete now. Not sure how to best approach this.

Reply:
Maybe a use for the `@Obsolete` annotation?